compress/compress.go

```diff
@@ -102,36 +156,126 @@ func New(algo CompressionAlgorithm, level CompressionLevel) (*Compressor, error)
+		if c.compressorZstdCgo != nil {
+			return c.compressorZstdCgo.Compress(src)
+		}
+		return c.compressorZstd.Compress(src)
+	})
 }
 
 func (c *Compressor) Decompress(src []byte) ([]byte, error) {
-	if c.compressorZstdCgo != nil {
-		return c.compressorZstdCgo.Decompress(src)
+	return c.withTimeout("decompress", src, func() ([]byte, error) {
+		if c.compressorZstdCgo != nil {
+			return c.compressorZstdCgo.Decompress(src)
+		}
+		return c.compressorZstd.Decompress(src)
+	})
+}
+
+// withTimeout wraps a compression/decompression operation with timeout protection.
+//
+// If no timeout is configured (c.settings.timeout == 0), the function executes directly
+// without any overhead. If a timeout is configured, the operation runs in a separate
+// goroutine with timeout monitoring.
+//
+// On timeout, this method returns an error (or panics if panicOnTimeout is enabled).
+// The goroutine executing the operation will continue running in the background and
+// cannot be cancelled due to lack of context support in the underlying compression
+// libraries. This is a known limitation - see WithTimeout for details.
+func (c *Compressor) withTimeout(operation string, data []byte, fn func() ([]byte, error)) ([]byte, error) {
+	// If no timeout configured, call function directly
+	if c.settings.timeout == 0 {
+		return fn()
+	}
+
+	type result struct {
+		data []byte
+		err  error
+	}
+
+	resultCh := make(chan result, 1)
+	go func() {
+		data, err := fn()
+		resultCh <- result{data: data, err: err}
+		close(resultCh)
+	}()
+
+	select {
+	case r := <-resultCh:
+		return r.data, r.err
+	case <-time.After(c.settings.timeout):
+		// Goroutine leak: the goroutine above will continue running until fn() completes.
+		// This is unavoidable without context cancellation support in the underlying libraries.
+
+		// Log timeout warning with data details
+		c.logTimeoutWarning(operation, data)
+
+		timeoutErr := fmt.Errorf("%s operation timeout after %v", operation, c.settings.timeout)
+		if c.settings.panicOnTimeout {
+			panic(timeoutErr)
+		}
+		return nil, timeoutErr
 	}
-	return c.compressorZstd.Decompress(src)
+}
+
+// logTimeoutWarning logs a warning when a timeout occurs, including safe serialization of the data.
+// The data is base64 encoded and truncated to the first 1KB to prevent excessive logging.
+func (c *Compressor) logTimeoutWarning(operation string, data []byte) {
+	const maxLoggedBytes = 1024
+
+	dataLen := len(data)
+	dataSample := data
+	isTruncated := false
+
+	if dataLen > maxLoggedBytes {
+		dataSample = data[:maxLoggedBytes]
+		isTruncated = true
+	}
+
+	encodedData := base64.StdEncoding.EncodeToString(dataSample)
+
+	c.settings.logger.Warnn(
+		"Compression operation timeout",
+		logger.NewStringField("operation", operation),
+		logger.NewDurationField("timeout", c.settings.timeout),
+		logger.NewIntField("dataLength", int64(dataLen)),
+		logger.NewBoolField("dataTruncated", isTruncated),
+		logger.NewStringField("dataBase64", encodedData),
```

Comment:
Why do we need to log the data here, aren't there any privacy concerns?

Reply:
We can enable/disable accordingly :+1: 
I can change it to `Debugn`.